### PR TITLE
[Merged by Bors] - feat: lemmas about monotonicity and asymptotics of `rpow`

### PR DESCRIPTION
**Pull request by dupuisf**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/SpecialFunctions/Exp.lean [hunk 1]
```diff
@@ -177,6 +177,11 @@ theorem tendsto_exp_atTop : Tendsto exp atTop atTop := by
   exact tendsto_atTop_mono' atTop B A
 #align real.tendsto_exp_at_top Real.tendsto_exp_atTop
 
+lemma _root_.Filter.Tendsto.exp_atTop {f : α → ℝ} (hf : Tendsto f l atTop) :
```
Comment (sgouezel) [inline Mathlib/Analysis/SpecialFunctions/Exp.lean:180]: I am not sure this one and the next one are really useful; aren't they just `Real.tendsto_exp_atTop.comp` and `Real.tendsto_exp_atBot.comp` (which is barely longer than calling this lemma by its name).

File: Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean [hunk 2]
```diff
@@ -51,6 +51,44 @@ theorem tendsto_rpow_neg_atTop {y : ℝ} (hy : 0 < y) : Tendsto (fun x : ℝ =>
...
+    Tendsto (rpow b) atBot (𝓝 (0:ℝ)) := by
+  show Tendsto (fun z => b^z) atBot (nhds 0)
+  simp_rw [Real.rpow_def_of_pos (by positivity : 0 < b)]
+  refine Tendsto.exp_atBot ?_
+  have h₁ : 0 < log b := by rw [log_pos_iff (by positivity)]; aesop
+  rw [tendsto_const_mul_atBot_of_pos h₁]
+  show atBot ≤ atBot
+  rfl
+
+lemma tendsto_rpow_atBot_of_base_lt_one (b : ℝ) (hb₀ : 0 < b) (hb₁ : b < 1) :
```
Comment (sgouezel) [inline Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean:74]: ditto

File: Mathlib/Analysis/SpecialFunctions/Pow/Real.lean [hunk 3]
```diff
@@ -634,6 +652,31 @@ theorem rpow_nat_inv_pow_nat {x : ℝ} (hx : 0 ≤ x) {n : ℕ} (hn : n ≠ 0) :
   rw [← rpow_nat_cast, ← rpow_mul hx, inv_mul_cancel hn0, rpow_one]
 #align real.rpow_nat_inv_pow_nat Real.rpow_nat_inv_pow_nat
 
+lemma rpow_strictMono_of_base_gt_one {b : ℝ} (hb : 1 < b) :
```
Comment (sgouezel) [inline Mathlib/Analysis/SpecialFunctions/Pow/Real.lean:655]:
```suggestion
lemma strictMono_rpow_of_one_lt_base {b : ℝ} (hb : 1 < b) :
```
to follow the order of symbols in the statement of the lemma. Same thing for the next lemmas.

File: Mathlib/Analysis/SpecialFunctions/Exp.lean (hunk 1 above)
Comment (dupuisf) [reply Mathlib/Analysis/SpecialFunctions/Exp.lean:180]: Yes, fair enough.

File: Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean [hunk 4]
```diff
@@ -51,6 +51,44 @@ theorem tendsto_rpow_neg_atTop {y : ℝ} (hy : 0 < y) : Tendsto (fun x : ℝ =>
     (tendsto_rpow_atTop hy).inv_tendsto_atTop
 #align tendsto_rpow_neg_at_top tendsto_rpow_neg_atTop
 
+lemma tendsto_rpow_atTop_of_base_lt_one (b : ℝ) (hb₀ : 0 < b) (hb₁ : b < 1) :
```
[the remark this answers is not in the document]
Comment (dupuisf) [reply Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean]: Generalized to `-1 < b`.

File: Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean (hunk 2 above)
Comment (dupuisf) [reply Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean]: This one is not true for `b ≤ 0`

Comment (sgouezel) [reply Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean]: Even for `b = 0`?

File: Mathlib/Analysis/SpecialFunctions/Pow/Real.lean (hunk 3 above)
Comment (dupuisf) [reply Mathlib/Analysis/SpecialFunctions/Pow/Real.lean]: Changed.

File: Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean [hunk 5]
```diff
@@ -51,6 +51,76 @@ theorem tendsto_rpow_neg_atTop {y : ℝ} (hy : 0 < y) : Tendsto (fun x : ℝ =>
     (tendsto_rpow_atTop hy).inv_tendsto_atTop
 #align tendsto_rpow_neg_at_top tendsto_rpow_neg_atTop
 
+open Asymptotics in
+lemma tendsto_rpow_atTop_of_base_lt_one (b : ℝ) (hb₀ : -1 < b) (hb₁ : b < 1) :
+    Tendsto (rpow b) atTop (𝓝 (0:ℝ)) := by
+  show Tendsto (fun z => b^z) atTop (nhds 0)
```
Comment (sgouezel) [inline Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean:57]:
```suggestion
  show Tendsto (fun z => b^z) atTop (𝓝 0)
```

File: Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean [hunk 6]
```diff
@@ -51,6 +51,76 @@ theorem tendsto_rpow_neg_atTop {y : ℝ} (hy : 0 < y) : Tendsto (fun x : ℝ =>
...
+  rcases le_or_gt b 0 with hb | hb
```
Comment (sgouezel) [inline Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean:58]: You should be able so streamline this with `rcases lt_trichotomy b 0 with hb|rfl|hb` which does directly the three cases distinction.

File: Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean [hunk 7]
```diff
@@ -51,6 +51,76 @@ theorem tendsto_rpow_neg_atTop {y : ℝ} (hy : 0 < y) : Tendsto (fun x : ℝ =>
...
+  case inl =>   -- b ≤ 0
+    simp_rw [Real.rpow_def_of_nonpos hb]
+    rcases lt_or_eq_of_le hb with hb | rfl
+    case inl =>  -- b < 0
+      have hb_ne : b ≠ 0 := by aesop
+      simp only [hb_ne, ite_false]
```
Comment (sgouezel) [inline Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean:64]:
```suggestion
      simp only [hb.ne, ite_false]
```
You don't need to resort to `aesop` here.

File: Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean [hunk 8]
```diff
@@ -51,6 +51,76 @@ theorem tendsto_rpow_neg_atTop {y : ℝ} (hy : 0 < y) : Tendsto (fun x : ℝ =>
...
+
+lemma tendsto_rpow_atTop_of_base_gt_one (b : ℝ) (hb : 1 < b) :
+    Tendsto (rpow b) atBot (𝓝 (0:ℝ)) := by
+  show Tendsto (fun z => b^z) atBot (nhds 0)
+  simp_rw [Real.rpow_def_of_pos (by positivity : 0 < b)]
+  refine tendsto_exp_atBot.comp ?_
+  have h₁ : 0 < log b := by rw [log_pos_iff (by positivity)]; aesop
+  rw [tendsto_const_mul_atBot_of_pos h₁]
+  show atBot ≤ atBot
+  rfl
```
Comment (sgouezel) [inline Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean:104]:
```suggestion
  exact tendsto_id
```

File: Mathlib/Analysis/SpecialFunctions/Pow/Real.lean [hunk 9]
```diff
@@ -477,6 +477,24 @@ theorem rpow_le_rpow_of_exponent_le (hx : 1 ≤ x) (hyz : y ≤ z) : x ^ y ≤ x
...
+theorem rpow_lt_rpow_of_exponent_neg {x y z : ℝ} (hy : 0 < y) (hxy : y < x) (hz : z < 0) :
+    x ^ z < y ^ z := by
+  have hx : 0 < x := hy.trans hxy
+  rw [←neg_neg z, Real.rpow_neg (le_of_lt hx) (-z), Real.rpow_neg (le_of_lt hy) (-z),
+      inv_lt_inv (rpow_pos_of_pos hx _) (rpow_pos_of_pos hy _)]
+  exact Real.rpow_lt_rpow (by positivity) hxy <| neg_pos_of_neg hz
+
+theorem rpow_le_rpow_of_exponent_nonpos {x y : ℝ} (hy : 0 < y) (hxy : y ≤ x) (hz : z ≤ 0) :
+    x ^ z ≤ y ^ z := by
+  rcases ne_or_eq z 0 with hz_zero | rfl
```
Comment (sgouezel) [inline Mathlib/Analysis/SpecialFunctions/Pow/Real.lean:489]: Here also you can use `lt_trichotomy`.

Comment (dupuisf) [reply Mathlib/Analysis/SpecialFunctions/Pow/Real.lean:489]: I don't think so: the two uses of `rcases` here involve different variables.

File: Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean [hunk 10]
```diff
@@ -51,6 +51,72 @@ theorem tendsto_rpow_neg_atTop {y : ℝ} (hy : 0 < y) : Tendsto (fun x : ℝ =>
...
+lemma tendsto_rpow_atTop_of_base_lt_one (b : ℝ) (hb₀ : -1 < b) (hb₁ : b < 1) :
+    Tendsto (rpow b) atTop (𝓝 (0:ℝ)) := by
+  show Tendsto (fun z => b^z) atTop (𝓝 0)
+  rcases lt_trichotomy b 0 with hb|rfl|hb
+  case inl =>   -- b < 0
+    simp_rw [Real.rpow_def_of_nonpos hb.le]
+    simp only [hb.ne, ite_false]
+    rw [←isLittleO_const_iff (c := (1:ℝ)) one_ne_zero]
+    have H : (1:ℝ) = 1 * 1 := by simp
+    rw [H]
```
Comment (j-loreaux) [inline Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean:64]:
```suggestion
    simp_rw [Real.rpow_def_of_nonpos hb.le, hb.ne, ite_false]
    rw [←isLittleO_const_iff (c := (1:ℝ)) one_ne_zero, (one_mul (1 : ℝ)).symm]
```

File: Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean [hunk 11]
```diff
@@ -51,6 +51,72 @@ theorem tendsto_rpow_neg_atTop {y : ℝ} (hy : 0 < y) : Tendsto (fun x : ℝ =>
...
+    case exp =>
+      rw [isLittleO_const_iff one_ne_zero]
+      have h₀ : log b = log (- -b) := by simp
+      rw [h₀, log_neg_eq_log]
+      have hb' : 0 < -b := by linarith
+      have h₁ : log (-b) < 0 := by rw [log_neg_iff hb']; linarith
+      refine tendsto_exp_atBot.comp ?_
+      rw [tendsto_const_mul_atBot_of_neg h₁]
+      show atTop ≤ atTop
+      rfl
```
Comment (j-loreaux) [inline Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean:75]:
```suggestion
      refine tendsto_exp_atBot.comp <| (tendsto_const_mul_atBot_of_neg ?_).mpr tendsto_id
      rw [←log_neg_eq_log, log_neg_iff (by linarith)]
      linarith
```

File: Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean [hunk 12]
```diff
@@ -51,6 +51,72 @@ theorem tendsto_rpow_neg_atTop {y : ℝ} (hy : 0 < y) : Tendsto (fun x : ℝ =>
...
+    case cos =>
+      rw [isBigO_iff]
+      exact ⟨1, eventually_of_forall fun x => by simp [Real.abs_cos_le_one]⟩
+  case inr.inl =>  -- b = 0
+    simp only [log_zero, zero_mul, exp_zero, one_mul, ite_true]
```
Comment (j-loreaux) [inline Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean:80]:
```suggestion
```

File: Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean [hunk 13]
```diff
@@ -51,6 +51,72 @@ theorem tendsto_rpow_neg_atTop {y : ℝ} (hy : 0 < y) : Tendsto (fun x : ℝ =>
...
+    rw [tendsto_pure]
+    filter_upwards [eventually_ne_atTop 0] with _ hx
+    simp [hx]
+  case inr.inr =>   -- b > 0
+    simp_rw [Real.rpow_def_of_pos hb]
+    have h₁ : log b < 0 := by rw [log_neg_iff hb]; exact hb₁
+    refine tendsto_exp_atBot.comp ?_
+    rw [tendsto_const_mul_atBot_of_neg h₁]
+    show atTop ≤ atTop
+    rfl
```
Comment (j-loreaux) [inline Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean:91]:
```suggestion
    refine tendsto_exp_atBot.comp <| (tendsto_const_mul_atBot_of_neg ?_).mpr tendsto_id
    exact (log_neg_iff hb).mpr hb₁
```

File: Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean [hunk 14]
```diff
@@ -51,6 +51,72 @@ theorem tendsto_rpow_neg_atTop {y : ℝ} (hy : 0 < y) : Tendsto (fun x : ℝ =>
...
+
+lemma tendsto_rpow_atTop_of_base_gt_one (b : ℝ) (hb : 1 < b) :
+    Tendsto (rpow b) atBot (𝓝 (0:ℝ)) := by
+  show Tendsto (fun z => b^z) atBot (nhds 0)
+  simp_rw [Real.rpow_def_of_pos (by positivity : 0 < b)]
+  refine tendsto_exp_atBot.comp ?_
+  have h₁ : 0 < log b := by rw [log_pos_iff (by positivity)]; aesop
+  rw [tendsto_const_mul_atBot_of_pos h₁]
+  exact tendsto_id
```
Comment (j-loreaux) [inline Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean:100]:
```suggestion
  refine tendsto_exp_atBot.comp <| (tendsto_const_mul_atBot_of_pos ?_).mpr tendsto_id
  exact (log_pos_iff (by positivity)).mpr <| by aesop
```

File: Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean [hunk 15]
```diff
@@ -51,6 +51,72 @@ theorem tendsto_rpow_neg_atTop {y : ℝ} (hy : 0 < y) : Tendsto (fun x : ℝ =>
...
+
+lemma tendsto_rpow_atBot_of_base_lt_one (b : ℝ) (hb₀ : 0 < b) (hb₁ : b < 1) :
+    Tendsto (rpow b) atBot atTop := by
+  show Tendsto (fun z => b^z) atBot atTop
+  simp_rw [Real.rpow_def_of_pos (by positivity : 0 < b)]
+  refine tendsto_exp_atTop.comp ?_
+  have h₁ : log b < 0 := by rw [log_neg_iff hb₀]; exact hb₁
+  rw [tendsto_const_mul_atTop_iff_neg (by show atBot ≤ atBot; simp)]
+  exact h₁
```
Comment (j-loreaux) [inline Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean:109]:
```suggestion
  refine tendsto_exp_atTop.comp <| (tendsto_const_mul_atTop_iff_neg <| tendsto_id (α := ℝ)).mpr ?_
  exact (log_neg_iff hb₀).mpr hb₁
```

File: Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean [hunk 16]
```diff
@@ -51,6 +51,72 @@ theorem tendsto_rpow_neg_atTop {y : ℝ} (hy : 0 < y) : Tendsto (fun x : ℝ =>
...
+
+lemma tendsto_rpow_atBot_of_base_gt_one (b : ℝ) (hb : 1 < b) : Tendsto (rpow b) atBot (𝓝 0) := by
+  show Tendsto (fun z => b^z) atBot (𝓝 0)
+  simp_rw [Real.rpow_def_of_pos (by positivity : 0 < b)]
+  refine tendsto_exp_atBot.comp ?_
+  have h₁ : 0 < log b := by rw [log_pos_iff (by positivity)]; aesop
+  rw [tendsto_const_mul_atBot_iff_pos (by show atBot ≤ atBot; simp)]
+  exact h₁
```
Comment (j-loreaux) [inline Mathlib/Analysis/SpecialFunctions/Pow/Asymptotics.lean:117]:
```suggestion
  refine tendsto_exp_atBot.comp <| (tendsto_const_mul_atBot_iff_pos <| tendsto_id (α := ℝ)).mpr ?_
  exact (log_pos_iff (by positivity)).mpr <| by aesop
```

**Review by j-loreaux — COMMENTED** [review]
These are just golfs, but I think they actually *improve* readability rather than hindering it. In particular, it put the main content in a single `refine` so that we can see these are all just the same pattern: `tendsto_exp_at{Bot,Top}.comp <| tendsto_const_mul_at{Bot,Top}_...`

This also avoids a lot of unnecessary `have`s and `show`s.

**Comment by sgouezel** [comment]
bors r+
Thanks!

**Pull request closed without merging**